File: src/core/truth-engine/store/append-only.ts
 /**
  * TRUTH ENGINE - APPEND-ONLY CORE STORE
  * 
  * Fundamental law: No delete. No update.
  * 
  * Allowed operations: CREATE, READ, SUPERSEDE
  * Forbidden operations: UPDATE, DELETE
  */
 
 import type { CoreObject, BaseClass } from '../core/ontology';
 import { enforceInvariants } from '../core/invariants';
 import { isValidId } from '../core/ids';
 
 /**
  * STORE STATE (in-memory for now, would be persistent in production)
  */
 interface StoreState {
   objects: Map<string, CoreObject>;
   writeLog: WriteLogEntry[];
 }
 
 interface WriteLogEntry {
   readonly id: string;
   readonly operation: 'CREATE' | 'SUPERSEDE';
   readonly timestamp: string;
   readonly writer: string;
   readonly object_hash: string;
 }
 
 const state: StoreState = {
   objects: new Map(),
   writeLog: [],
 };
 
 /**
  * ALLOWED OPERATIONS
  */
 type AllowedOperation = 'CREATE' | 'READ' | 'SUPERSEDE';
 
 /**
  * WRITE RESULT
  */
 export interface WriteResult {
   readonly success: boolean;
   readonly operation: AllowedOperation;
   readonly id: string;
   readonly timestamp: string;
   readonly error?: string;
 }
 
 /**
  * CREATE - Write new object (only if ID doesn't exist)
  */
 export function create(obj: CoreObject, writer: string): WriteResult {
   const timestamp = new Date().toISOString();
   
   // Enforce all invariants BEFORE write
   try {
     enforceInvariants(obj);
   } catch (error) {
     return {
       success: false,
       operation: 'CREATE',
       id: obj.id,
       timestamp,
       error: error instanceof Error ? error.message : 'Unknown invariant violation',
     };
   }
   
   // Check for duplicate ID (no overwrites)
   if (state.objects.has(obj.id)) {
     return {
       success: false,
       operation: 'CREATE',
       id: obj.id,
       timestamp,
       error: 'FORBIDDEN: Object with this ID already exists. Use SUPERSEDE instead.',
     };
   }
   
   // Validate ID format
   if (!isValidId(obj.id)) {
     return {
       success: false,
       operation: 'CREATE',
       id: obj.id,
       timestamp,
       error: 'INVALID: ID format is incorrect',
     };
   }
   
   // Write to store
   state.objects.set(obj.id, Object.freeze(obj)); // Immutable
   
   // Log the write
   state.writeLog.push({
     id: obj.id,
     operation: 'CREATE',
     timestamp,
     writer,
     object_hash: hashObject(obj),
   });
   
   return {
     success: true,
     operation: 'CREATE',
     id: obj.id,
     timestamp,
   };
 }
 
 /**
  * READ - Get object by ID
  */
 export function read(id: string): CoreObject | null {
   return state.objects.get(id) || null;
 }
 
 /**
  * READ ALL - Get all objects of a type
  */
 export function readAll(baseClass?: BaseClass): CoreObject[] {
   const all = Array.from(state.objects.values());
   if (!baseClass) return all;
   return all.filter(obj => obj.baseClass === baseClass);
 }
 
 /**
  * EXISTS - Check if ID exists
  */
 export function exists(id: string): boolean {
   return state.objects.has(id);
 }
 
 /**
  * GET WRITE LOG - Audit trail
  */
 export function getWriteLog(): readonly WriteLogEntry[] {
   return Object.freeze([...state.writeLog]);
 }
 
 /**
  * FORBIDDEN OPERATIONS - These throw immediately
  */
 export function update(): never {
   throw new Error('FORBIDDEN: UPDATE operation is not permitted. Truth Engine is append-only.');
 }
 
 export function deleteObject(): never {
   throw new Error('FORBIDDEN: DELETE operation is not permitted. Truth Engine is append-only.');
 }
 
 /**
  * HASH OBJECT (for audit log)
  */
 function hashObject(obj: CoreObject): string {
   const str = JSON.stringify(obj, Object.keys(obj).sort());
   let hash = 0;
   for (let i = 0; i < str.length; i++) {
     const char = str.charCodeAt(i);
     hash = ((hash << 5) - hash) + char;
     hash = hash & hash;
   }
   return Math.abs(hash).toString(16).padStart(8, '0');
 }
 
 /**
  * STORE STATS
  */
 export function getStats(): {
   totalObjects: number;
   objectsByClass: Record<string, number>;
   totalWrites: number;
 } {
   const objects = Array.from(state.objects.values());
   const byClass: Record<string, number> = {};
   
   for (const obj of objects) {
     byClass[obj.baseClass] = (byClass[obj.baseClass] || 0) + 1;
   }
   
   return {
     totalObjects: objects.length,
     objectsByClass: byClass,
     totalWrites: state.writeLog.length,
   };
 }
 
 /**
  * CLEAR STORE (for testing only - would not exist in production)
  */
 export function _clearForTesting(): void {
   state.objects.clear();
   state.writeLog.length = 0;
 }

File: src/core/truth-engine/python-reference/append_only.py
 """
 TRUTH ENGINE - APPEND-ONLY CORE STORE
 
 Fundamental law: No delete. No update.
 
 Allowed operations: CREATE, READ, SUPERSEDE
 Forbidden operations: UPDATE, DELETE
 """
 
 import hashlib
 import json
 from datetime import datetime
 from typing import List, Dict, Any, Optional
 from dataclasses import dataclass, field
 
 from invariants import enforce_invariants
 from ids import is_valid_id
 
 
 @dataclass
 class WriteLogEntry:
     id: str
     operation: str  # 'CREATE' or 'SUPERSEDE'
     timestamp: str
     writer: str
     object_hash: str
 
 
 @dataclass
 class WriteResult:
     success: bool
     operation: str
     id: str
     timestamp: str
     error: Optional[str] = None
 
 
 class AppendOnlyStore:
     """
     Append-only truth store.
     
     NO DELETE. NO UPDATE. EVER.
     """
     
     def __init__(self):
         self._records: Dict[str, Dict[str, Any]] = {}
         self._write_log: List[WriteLogEntry] = []
     
     def _hash_object(self, obj: Dict[str, Any]) -> str:
         """Generate hash for audit log."""
         canonical = json.dumps(obj, sort_keys=True)
         return hashlib.sha256(canonical.encode()).hexdigest()[:12]
     
     def create(self, record: Dict[str, Any], writer: str) -> WriteResult:
         """
         CREATE - Write new object (only if ID doesn't exist).
         """
         timestamp = datetime.utcnow().isoformat() + "Z"
         record_id = record.get("id", "")
         
         # Enforce all invariants BEFORE write
         try:
             enforce_invariants(record)
         except ValueError as e:
             return WriteResult(
                 success=False,
                 operation="CREATE",
                 id=record_id,
                 timestamp=timestamp,
                 error=str(e)
             )
         
         # Check for duplicate ID (no overwrites)
         if record_id in self._records:
             return WriteResult(
                 success=False,
                 operation="CREATE",
                 id=record_id,
                 timestamp=timestamp,
                 error="FORBIDDEN: Object with this ID already exists. Use SUPERSEDE instead."
             )
         
         # Validate ID format
         if not is_valid_id(record_id):
             return WriteResult(
                 success=False,
                 operation="CREATE",
                 id=record_id,
                 timestamp=timestamp,
                 error="INVALID: ID format is incorrect"
             )
         
         # Write to store (immutable copy)
         self._records[record_id] = dict(record)
         
         # Log the write
         self._write_log.append(WriteLogEntry(
             id=record_id,
             operation="CREATE",
             timestamp=timestamp,
             writer=writer,
             object_hash=self._hash_object(record)
         ))
         
         return WriteResult(
             success=True,
             operation="CREATE",
             id=record_id,
             timestamp=timestamp
         )
     
     def read(self, id: str) -> Optional[Dict[str, Any]]:
         """READ - Get object by ID."""
         return self._records.get(id)
     
     def read_all(self, base_class: Optional[str] = None) -> List[Dict[str, Any]]:
         """READ ALL - Get all objects, optionally filtered by type."""
         all_records = list(self._records.values())
         if base_class:
             return [r for r in all_records if r.get("base_class") == base_class]
         return all_records
     
     def exists(self, id: str) -> bool:
         """EXISTS - Check if ID exists."""
         return id in self._records
     
     def get_write_log(self) -> List[WriteLogEntry]:
         """GET WRITE LOG - Audit trail."""
         return list(self._write_log)
     
     def update(self) -> None:
         """FORBIDDEN: UPDATE operation is not permitted."""
         raise PermissionError(
             "FORBIDDEN: UPDATE operation is not permitted. "
             "Truth Engine is append-only."
         )
     
     def delete(self) -> None:
         """FORBIDDEN: DELETE operation is not permitted."""
         raise PermissionError(
             "FORBIDDEN: DELETE operation is not permitted. "
             "Truth Engine is append-only."
         )
     
     def stats(self) -> Dict[str, Any]:
         """STORE STATS"""
         by_class: Dict[str, int] = {}
         for record in self._records.values():
             bc = record.get("base_class", "unknown")
             by_class[bc] = by_class.get(bc, 0) + 1
         
         return {
             "total_objects": len(self._records),
             "objects_by_class": by_class,
             "total_writes": len(self._write_log)
         }